aten/src/ATen/native/TensorAdvancedIndexing.cpp

```diff
@@ -1261,6 +1283,54 @@ TORCH_IMPL_FUNC(scatter_add)
   }
 }
 
+TORCH_IMPL_FUNC(_scatter_reduce_structured_cpu)
+(const Tensor& self,
+ int64_t dim,
+ const Tensor& index,
+ const c10::string_view reduce,
+ const c10::optional<int64_t> output_size,
+ const Tensor& out) {
+
+  TORCH_CHECK(self.dim() == index.dim(),
+      "Shape mismatch between `self` (got ", self.sizes(), ") and `index` (got ", index.sizes(), ")");
+  for (int64_t i = 0; i < self.dim(); i++) {
+    TORCH_CHECK(self.size(i) == index.size(i),
+        "Shape mismatch between `self` (got ", self.sizes(), ") and `index` (got ", index.sizes(), ")");
+  }
+
+  TORCH_CHECK(reduce == "sum", "`reduce` argument must be 'sum'");
+
+  dim = dim < 0 ? dim + self.dim() : dim;
+
+  out.zero_();
+
+  AT_DISPATCH_ALL_TYPES_AND2(kHalf, kBFloat16, self.scalar_type(), "_scatter_reduce", [&] {
+    auto self_data = self.data_ptr<scalar_t>();
+    auto index_data = index.data_ptr<int64_t>();
+    auto out_data = out.data_ptr<scalar_t>();
+
+    int64_t offset1 = 1, offset2 = 1;
+    for (int64_t d = 0; d < dim; d++)
+      offset1 *= self.size(d);
```

Comment:
this would work only for contiguous tensors?

Reply:
This PR is WIP for now.

Reply:
@rusty1s - we should probably add a check that the inputs that are used here are contiguous